src/main/kotlin/io/libp2p/core/crypto/Key.kt

```diff
@@ -0,0 +1,261 @@
+import io.libp2p.core.crypto.keys.unmarshalEd25519PublicKey
+import io.libp2p.core.crypto.keys.unmarshalRsaPrivateKey
+import io.libp2p.core.crypto.keys.unmarshalRsaPublicKey
+import io.libp2p.core.crypto.keys.unmarshalSecp256k1PrivateKey
+import io.libp2p.core.crypto.keys.unmarshalSecp256k1PublicKey
+import io.libp2p.core.types.toHex
+import org.bouncycastle.crypto.digests.SHA256Digest
+import org.bouncycastle.crypto.digests.SHA512Digest
+import org.bouncycastle.crypto.macs.HMac
+import org.bouncycastle.crypto.params.KeyParameter
+import crypto.pb.Crypto.PrivateKey as PbPrivateKey
+import crypto.pb.Crypto.PublicKey as PbPublicKey
+
+enum class KEY_TYPE {
+
+    /**
+     * RSA is an enum for the supported RSA key type
+     */
+    RSA,
+
+    /**
+     * Ed25519 is an enum for the supported Ed25519 key type
+     */
+    ED25519,
+
+    /**
+     * Secp256k1 is an enum for the supported Secp256k1 key type
+     */
+    SECP256K1,
+
+    /**
+     * ECDSA is an enum for the supported ECDSA key type
+     */
+    ECDSA
+}
+
+interface Key {
+
+    val keyType: crypto.pb.Crypto.KeyType
+
+    /**
+     * Bytes returns a serialized, storeable representation of this key.
+     */
+    fun bytes(): ByteArray
+
+    fun raw(): ByteArray
+}
+
+/**
+ * PrivKey represents a private key that can be used to generate a public key,
+ * sign data, and decrypt data that was encrypted with a public key.
+ */
+abstract class PrivKey(override val keyType: Crypto.KeyType) : Key {
+
+    /**
+     * Cryptographically sign the given bytes.
+     */
+    abstract fun sign(data: ByteArray): ByteArray
+
+    /**
+     * Return a public key paired with this private key.
+     */
+    abstract fun publicKey(): PubKey
+
+    override fun bytes(): ByteArray = marshalPrivateKey(this)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+        return bytes().contentEquals((other as PrivKey).bytes())
+    }
+}
+
+/**
+ * PubKey is a public key.
+ */
+abstract class PubKey(override val keyType: Crypto.KeyType) : Key {
+
+    /**
+     * Verify that 'sig' is the signed hash of 'data'.
+     */
+    abstract fun verify(data: ByteArray, signature: ByteArray): Boolean
+
+    override fun bytes(): ByteArray = marshalPublicKey(this)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+        return bytes().contentEquals((other as PubKey).bytes())
+    }
+}
+
+class BadKeyTypeException : Exception("Invalid or unsupported key type")
+
+/**
+ * Generates a new key pair of the provided type.
+ * @param type the type key to be generated.
+ * @param bits the number of bits desired for the key (only applicable for RSA).
+ */
+fun generateKeyPair(type: KEY_TYPE, bits: Int = 0): Pair<PrivKey, PubKey> {
+
+    return when (type) {
+        KEY_TYPE.RSA -> generateRsaKeyPair(bits)
```

Comment:
We might want to use polymorphism here rather than a finite non-extendable set, but not important now.

Reply:
Just borrowed this from Web3. Probably you are right. Didn't pay too much attention here